// @amplitude/analytics-browser/lib/cjs/video-capture/video-capture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackVideo = exports.VideoCapture = void 0;
var tslib_1 = require("tslib");
var analytics_core_1 = require("@amplitude/analytics-core");
var VideoCapture = /** @class */ (function () {
    function VideoCapture(amplitude) {
        this.amplitude = amplitude;
        this.videoEl = null;
        this.embeddedVideoPlayer = null;
        this.extraEventProperties = {};
        this.listeners = [];
        this.onRemoveListeners = [];
    }
    /**
     * Specify a video element to capture events from
     *
     * @param videoEl - The HTML video element to capture events from.
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.withVideoElement = function (videoEl) {
        this.videoEl = videoEl;
        return this;
    };
    /**
     * Specify an embedded video player.js instance to capture events from
     * @param player - The embedded video player.js instance to capture events from.
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.withEmbeddedPlayer = function (player) {
        this.embeddedVideoPlayer = player;
        return this;
    };
    /**
     * Specify a vendor to capture extra vendor-specific event properties
     *
     * @param vendor - The vendor of the video player. Currently only "mux" is supported.
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.withVendor = function (vendor) {
        this.vendor = vendor;
        return this;
    };
    /**
     * Specify extra event properties to include in all captured events
     *
     * @param properties - The extra event properties to include in the Amplitude event.
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.withExtraEventProperties = function (properties) {
        this.extraEventProperties = properties;
        return this;
    };
    /**
     * Track a "Video Content Started" event every time the video starts playing
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.captureVideoStarted = function () {
        var _this = this;
        this.listeners.push(function (previousState, nextState) {
            if (previousState.playbackState !== 'playing' && nextState.playbackState === 'playing') {
                // TODO: placeholder for Heartbeat Start Event
                _this.amplitude.track('Video Content Started', tslib_1.__assign(tslib_1.__assign({}, nextState.lastEvent), _this.extraEventProperties));
            }
        });
        return this;
    };
    /**
     * Track a "Video Content Stopped" event every time the video stops playing
     * @returns The VideoCapture instance.
     */
    VideoCapture.prototype.captureVideoStopped = function () {
        var _this = this;
        this.listeners.push(function (previousState, nextState) {
            if (previousState.playbackState === 'playing' && nextState.playbackState !== 'playing') {
                // placeholder for Heartbeat Stop Event
                _this.amplitude.track('Video Content Stopped', tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, nextState.lastEvent), { watch_duration: nextState.watchTime }), _this.extraEventProperties));
            }
        });
        return this;
    };
    // Placeholder: may need a generic state change listener to capture unusual events or to have
    // more control over the event tracking.
    // withStateChangeListener(listener: (previousState: VideoState, nextState: VideoState) => void): VideoCapture {
    /**
     * Start capturing analytics events for the video element
     * @returns The VideoCapture instance.
     * @throws An error if the video element is not specified.
     */
    VideoCapture.prototype.start = function () {
        var _this = this;
        var _a;
        var videoEl = (_a = this.videoEl) !== null && _a !== void 0 ? _a : this.embeddedVideoPlayer;
        if (!videoEl) {
            throw new Error('Video element not specified. Use withVideoElement() or withEmbeddedPlayer() to specify the video element.');
        }
        if (this.videoEl && this.embeddedVideoPlayer) {
            throw new Error('Both video element and embedded video player specified. Use only one of withVideoElement() or withEmbeddedPlayer() to specify the video element.');
        }
        var videoObserver = new analytics_core_1.VideoObserver({
            videoEl: videoEl,
            onStateChange: function (previousState, nextState) {
                _this.listeners.forEach(function (listener) { return listener(previousState, nextState); });
            },
            vendor: this.vendor,
            isEmbedded: !!this.embeddedVideoPlayer,
        });
        this.onRemoveListeners.push(function () {
            videoObserver.destroy();
        });
        return this;
    };
    VideoCapture.prototype.stop = function () {
        this.onRemoveListeners.forEach(function (listener) { return listener(); });
        this.onRemoveListeners = [];
    };
    return VideoCapture;
}());
exports.VideoCapture = VideoCapture;
/**
 * Track video analytics events for an HTML video element or embedded video player.js instance.
 *
 * Captures Video Started and Video Stopped events.
 *
 * @experimental This function is experimental and may not be stable.
 * @param amplitude - The Amplitude client instance.
 * @param videoEl - The HTML video element or embedded video player.js instance to capture events from.
 * @param options - The options for the video capture.
 * @returns A function to stop the video capture.
 */
function trackVideo(amplitude, videoEl, options) {
    var _a;
    if (options === void 0) { options = {}; }
    var videoCapture = new VideoCapture(amplitude);
    if (videoEl instanceof HTMLVideoElement) {
        videoCapture.withVideoElement(videoEl);
    }
    else {
        videoCapture.withEmbeddedPlayer(videoEl);
    }
    if (options.vendor) {
        videoCapture.withVendor(options.vendor);
    }
    var extraEventProperties = (_a = options.extraEventProperties) !== null && _a !== void 0 ? _a : {};
    try {
        videoCapture
            .withExtraEventProperties(tslib_1.__assign({ view_session_id: (0, analytics_core_1.UUID)() }, extraEventProperties))
            .captureVideoStarted()
            .captureVideoStopped()
            .start();
    }
    catch (error) {
        return error;
    }
    return function () { return videoCapture.stop(); };
}
exports.trackVideo = trackVideo;
