// @amplitude/analytics-browser/lib/esm/transports/fetch.js
import { __assign, __awaiter, __extends, __generator } from "tslib";
import { BaseTransport, compressToGzipArrayBuffer, isCompressionStreamAvailable, MIN_GZIP_UPLOAD_BODY_SIZE_BYTES, } from '@amplitude/analytics-core';
// Conservative 1/4 of the shared 64 KiB keepalive budget. Why 16K: https://github.com/amplitude/Amplitude-TypeScript/pull/1781
export var KEEPALIVE_MAX_BODY_SIZE_BYTES = 16 * 1024;
// Temporary browser-specific fetch transport with gzip support.
// TODO: Merge this implementation back into @amplitude/analytics-core FetchTransport
// once React Native SDK supports request body gzip.
var FetchTransport = /** @class */ (function (_super) {
    __extends(FetchTransport, _super);
    function FetchTransport(customHeaders, enableKeepalive, referrerPolicy) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var _this = _super.call(this) || this;
        _this.customHeaders = customHeaders;
        // Enabled unless explicitly disabled.
        _this.enableKeepalive = enableKeepalive !== false;
        // Enable this to control Referer header behavior for fetch requests.
        // Otherwise, leave it as the default browser behavior.
        _this.referrerPolicy = referrerPolicy;
        return _this;
    }
    FetchTransport.prototype.send = function (serverUrl, payload, shouldCompressUploadBody) {
        if (shouldCompressUploadBody === void 0) { shouldCompressUploadBody = false; }
        return __awaiter(this, void 0, void 0, function () {
            var bodyString, shouldCompressBody, body, headers, compressed, bodySize, options, response, responseText;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        /* istanbul ignore if */
                        if (typeof fetch === 'undefined') {
                            throw new Error('FetchTransport is not supported');
                        }
                        bodyString = JSON.stringify(payload);
                        shouldCompressBody = shouldCompressUploadBody &&
                            bodyString.length >= MIN_GZIP_UPLOAD_BODY_SIZE_BYTES &&
                            isCompressionStreamAvailable();
                        body = bodyString;
                        headers = {
                            'Content-Type': 'application/json',
                            Accept: '*/*',
                        };
                        if (!shouldCompressBody) return [3 /*break*/, 2];
                        return [4 /*yield*/, compressToGzipArrayBuffer(bodyString)];
                    case 1:
                        compressed = _a.sent();
                        if (compressed) {
                            headers['Content-Encoding'] = 'gzip';
                            body = compressed;
                        }
                        _a.label = 2;
                    case 2:
                        headers = __assign(__assign({}, this.customHeaders), headers);
                        bodySize = typeof body === 'string' ? body.length : body.byteLength;
                        options = {
                            headers: headers,
                            body: body,
                            method: 'POST',
                            keepalive: this.enableKeepalive && bodySize <= KEEPALIVE_MAX_BODY_SIZE_BYTES,
                        };
                        if (this.referrerPolicy) {
                            options.referrerPolicy = this.referrerPolicy;
                        }
                        return [4 /*yield*/, fetch(serverUrl, options)];
                    case 3:
                        response = _a.sent();
                        return [4 /*yield*/, response.text()];
                    case 4:
                        responseText = _a.sent();
                        try {
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                            return [2 /*return*/, this.buildResponse(JSON.parse(responseText))];
                        }
                        catch (_b) {
                            return [2 /*return*/, this.buildResponse({ code: response.status })];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return FetchTransport;
}(BaseTransport));
export { FetchTransport };
