// @amplitude/plugin-autocapture-browser/lib/esm/autocapture/track-long-task.js
import { __assign, __values } from "tslib";
import { AMPLITUDE_MAIN_THREAD_BLOCK_EVENT } from '../constants';
import { isUrlAllowed } from '../helpers';
var DEFAULT_DURATION_THRESHOLD = 100; // ms
var MEASURE_BUFFER_WINDOW_MS = 10000;
function getOverlappingMeasures(entry, measures) {
    var taskEnd = entry.startTime + entry.duration;
    return measures
        .filter(function (measure) { return measure.startTime < taskEnd && measure.startTime + measure.duration > entry.startTime; })
        .map(function (measure) { return measure.name; });
}
function buildLoAFProperties(entry, measures) {
    var _a;
    var overlappingMeasures = getOverlappingMeasures(entry, measures);
    var scripts = (_a = entry.scripts) !== null && _a !== void 0 ? _a : [];
    var scriptURLs = scripts.map(function (s) { return s.sourceURL; }).filter(Boolean);
    var scriptFunctions = scripts.map(function (s) { return s.sourceFunctionName; }).filter(Boolean);
    var scriptPositions = scripts
        .map(function (s) { return s.sourceCharPosition; })
        .filter(function (p) { return typeof p === 'number' && p >= 0; });
    var invokerTypes = scripts.map(function (s) { return s.invokerType; }).filter(Boolean);
    var invokers = scripts.map(function (s) { return s.invoker; }).filter(Boolean);
    return __assign(__assign(__assign(__assign(__assign(__assign(__assign({ '[Amplitude] Main Thread Block Source': 'long-animation-frame', '[Amplitude] Main Thread Block Duration': entry.duration, '[Amplitude] Main Thread Block Blocking Duration': entry.blockingDuration, '[Amplitude] Main Thread Block Start Time': entry.startTime }, (overlappingMeasures.length > 0 && { '[Amplitude] Main Thread Block Measures': overlappingMeasures })), { '[Amplitude] Main Thread Block Render Start': entry.renderStart, '[Amplitude] Main Thread Block Style And Layout Start': entry.styleAndLayoutStart, '[Amplitude] Main Thread Block Script Count': scripts.length }), (scriptURLs.length > 0 && { '[Amplitude] Main Thread Block Script URLs': scriptURLs })), (scriptFunctions.length > 0 && { '[Amplitude] Main Thread Block Script Functions': scriptFunctions })), (scriptPositions.length > 0 && { '[Amplitude] Main Thread Block Script Positions': scriptPositions })), (invokerTypes.length > 0 && { '[Amplitude] Main Thread Block Invoker Types': invokerTypes })), (invokers.length > 0 && { '[Amplitude] Main Thread Block Invokers': invokers }));
}
function buildLongTaskProperties(entry, measures) {
    var _a;
    var overlappingMeasures = getOverlappingMeasures(entry, measures);
    var attribution = (_a = entry.attribution) !== null && _a !== void 0 ? _a : [];
    return __assign(__assign({ '[Amplitude] Main Thread Block Source': 'long-task', '[Amplitude] Main Thread Block Duration': entry.duration, '[Amplitude] Main Thread Block Blocking Duration': entry.duration, '[Amplitude] Main Thread Block Start Time': entry.startTime }, (overlappingMeasures.length > 0 && { '[Amplitude] Main Thread Block Measures': overlappingMeasures })), (attribution.length > 0 && {
        '[Amplitude] Main Thread Block Attribution': attribution.map(function (a) { return a.name; }),
    }));
}
function getSupportedEntryType() {
    /* istanbul ignore next */
    if (typeof PerformanceObserver === 'undefined')
        return null;
    try {
        var supported = PerformanceObserver.supportedEntryTypes;
        if (supported.includes('long-animation-frame'))
            return 'long-animation-frame';
        if (supported.includes('longtask'))
            return 'longtask';
    }
    catch (_a) {
        // ignore
    }
    return null;
}
export function trackMainThreadBlock(_a) {
    var amplitude = _a.amplitude, options = _a.options, _b = _a.durationThreshold, durationThreshold = _b === void 0 ? DEFAULT_DURATION_THRESHOLD : _b;
    var entryType = getSupportedEntryType();
    /* istanbul ignore next */
    if (!entryType) {
        return { unsubscribe: function () { return void 0; } };
    }
    var measures = [];
    var measureObserver = new PerformanceObserver(function (list) {
        var e_1, _a;
        var now = performance.now();
        try {
            for (var _b = __values(list.getEntries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var entry = _c.value;
                measures.push(entry);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var cutoff = now - MEASURE_BUFFER_WINDOW_MS;
        while (measures.length > 0 && measures[0].startTime < cutoff) {
            measures.shift();
        }
    });
    try {
        measureObserver.observe({ entryTypes: ['measure'] });
    }
    catch (_c) {
        // measure not supported — continue without it
    }
    var blockObserver = new PerformanceObserver(function (list) {
        var e_2, _a;
        try {
            for (var _b = __values(list.getEntries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var entry = _c.value;
                if (!isUrlAllowed(options)) {
                    return;
                }
                if (entry.duration < durationThreshold) {
                    continue;
                }
                var properties = entryType === 'long-animation-frame'
                    ? buildLoAFProperties(entry, measures)
                    : buildLongTaskProperties(entry, measures);
                amplitude.track(AMPLITUDE_MAIN_THREAD_BLOCK_EVENT, properties);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    });
    try {
        blockObserver.observe({ entryTypes: [entryType] });
    }
    catch (_d) {
        measureObserver.disconnect();
        return { unsubscribe: function () { return void 0; } };
    }
    return {
        unsubscribe: function () {
            blockObserver.disconnect();
            measureObserver.disconnect();
        },
    };
}
